File: backend/src/merchant/api/MerchantStoreApi.scala
package delivery.merchant.api

import cats.effect.IO
import delivery.shared.api.ApiPlan
import delivery.merchant.objects.{CreateStoreRequest, CreateStoreResponse}
import delivery.merchant.state.MerchantDomainOps
import delivery.shared.objects.DeliveryState
import delivery.shared.state.DeliveryStateOps
import org.typelevel.log4cats.slf4j.Slf4jLogger

object MerchantStoreApi extends ApiPlan[MerchantStoreApi.MerchantStoreCommand, Either[String, MerchantStoreApi.MerchantStoreSuccess]]:

  final case class MerchantStoreCommand(
      state: DeliveryState,
      username: String,
      body: CreateStoreRequest
  )

  private val logger = Slf4jLogger.getLogger[IO]

  override val name: String = "MerchantStoreApi"

  override def plan(input: MerchantStoreApi.MerchantStoreCommand): IO[Either[String, MerchantStoreApi.MerchantStoreSuccess]] =
    for
      _ <- logger.info(s"$name started, username=${input.username}")
      response <- MerchantDomainOps.createMerchantStore(input.state.merchant, input.username, input.body.storeName, input.body.address).flatMap {
        case Left(msg) => IO.pure(Left(msg))
        case Right((nextMerchant, merchantId)) =>
          IO.pure(
            Right(MerchantStoreSuccess(DeliveryStateOps.withMerchantState(input.state, nextMerchant), CreateStoreResponse(ok = true, merchantId = merchantId)))
          )
      }
      _ <- logger.info(s"$name finished, success=${response.isRight}")
    yield response

  final case class MerchantStoreSuccess(nextState: DeliveryState, response: CreateStoreResponse)

end MerchantStoreApi

File: frontend/src/api/merchant/MerchantStoreApi.ts
import { APIMessage } from '@/api/shared/APIMessage'
import type { TaskIO } from '@/api/shared/TaskIO'
import { sendAPI } from '@/api/shared/sendAPI'
import type { CreateStoreRequest } from '@/objects/merchant/CreateStoreRequest'

class MerchantStoreAPI extends APIMessage<string> {
  readonly apiName = 'merchantstoreapi'
  readonly storeName: string
  readonly address: string

  constructor(storeName: string, address: string) {
    super()
    this.storeName = storeName
    this.address = address
  }
}

export function createMerchantStoreIO(input: CreateStoreRequest): TaskIO<string> {
  return sendAPI(new MerchantStoreAPI(input.storeName, input.address))
}
